rest-api/dao/biobank_order_dao.py

```diff
@@ -339,4 +391,140 @@ def to_client_json(self, model):
+      client_json['restoredUsername'] = model.restoredUsername
+      client_json['version'] = model.version
+
+    if model.orderStatus == BiobankOrderStatus.AMENDED:
+      client_json['amendedReason'] = model.amendedReason
+      client_json['amendedSiteId'] = model.amendedSiteId
+      client_json['amendedUsername'] = model.amendedUsername
+      client_json['version'] = model.version
+
     return client_json
+
+  def _do_update(self, session, order, resource):
+    order.lastModified = clock.CLOCK.now()
+    order.biobankOrderId = resource.biobankOrderId
+    order.orderStatus = BiobankOrderStatus.AMENDED
+    order.amendedTime = clock.CLOCK.now()
+    order.logPosition = LogPosition()
+    order.version += 1
+    # Ensure that if an order was previously cancelled/restored those columns are removed.
+    self._clear_cancelled_and_restored_fields(order)
+
+    super(BiobankOrderDao, self)._do_update(session, order, resource)
+    session.add(order.logPosition)
+
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+
+  def update_with_patch(self, id_, resource, expected_version):
+    """creates an atomic patch request on an object. It will fail if the object
+    doesn't exist already, or if obj.version does not match the version of the existing object.
+    May modify the passed in object."""
+    with self.session() as session:
+      obj = self.get_with_children_in_session(session, id_, for_update=True)
+      return self.patch_update_with_session(session, obj, resource, expected_version)
+
+  def _do_update_with_patch(self, session, order, resource):
+    order.lastModified = clock.CLOCK.now()
+    order.logPosition = LogPosition()
+    order.version += 1
+    if resource['status'].lower() == 'cancelled':
+      order.amendedReason = resource['amendedReason']
+      order.cancelledUsername = resource['cancelledInfo']['author']['value']
+      order.cancelledSiteId = get_site(resource['cancelledInfo'])
+      order.cancelledTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.CANCELLED
+    elif resource['status'].lower() == 'restored':
+      order.amendedReason = resource['amendedReason']
+      order.restoredUsername = resource['restoredInfo']['author']['value']
+      order.restoredSiteId = get_site(resource['restoredInfo'])
+      order.restoredTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.UNSET
+    else:
+      raise BadRequest('status must be restored or cancelled for patch request.')
+
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+    super(BiobankOrderDao, self)._do_update(session, order, resource)
+    return order
+
+  def _validate_patch_update(self, session, model, resource, expected_version):
+    required_cancelled_fields = ['amendedReason', 'cancelledInfo', 'status']
```

Comment:
We could make a _FhirBiobankOrderPatch class that is used for parsing this stuff, like _FhirBiobankOrder... I don't really mind either way... just noting.